lib/using_restricted_array.rb

```diff
@@ -6,87 +6,144 @@
+  i = 0
+  while array[i] != nil
+    puts "#{array[i]}"
+    i += 1
+  end
+  
+  # For an unsorted array, searches for 'value_to_find'.
+  # Returns true if found, false otherwise.
+  # Time complexity: OofN linear bc the loop only goes the length
+  # Space complexity: Oof1 constant bc it onky returns true or false
+  def search(array, length, value_to_find)
+    length.times do |i|
+      if array[i] == value_to_find
+        return true
       end
-      temp_index += 1 # move to next index
-    end
-    if min_index != index # next minimum value is not at current index, swap
-      temp = array[min_index]
-      array[min_index] = array[index]
-      array[index] = temp
+    else i += 1
     end
   end
-end
-## --- END OF METHODS ---
+  
+  # Finds and returns the largest integer value the array
+  # Assumes that the array is not sorted.
+  # Time complexity: OofN linear/while loop only goes the length of the array
+  # Space complexity: Oof1 constant bc were only returning the single largest
+  def find_largest(array, length)
+    kaboom = 999
+    i = 0
+    largest = array[i]
+    while i + 1 < length && 
+      array[i + 1] != kaboom
+      index += 1
+      if array[i] > largest
+      end
+      return largest
+    end
+    
+    # Finds and returns the smallest integer value in the array
+    # Assumes that the array is not sorted.
+    # Time complexity: OofN linear bc the loop only goes the length of the array
+    # Space complexity: Oof1 bc size(memory) isnt changing [smallest, i]
+    def find_smallest(array, length)
+      i = 0
+      smallest = array[i]
+      while i - 1 > length && 
+        array[i - 1] != kaboom
+        index += 1
+        if array[i] < smallest
+        end
+        return smallest
+      end
+      
+      # Reverses the values in the integer array in place
+      # Time complexity: OofN - n is number of elements in array
+      # Space complexity: Oof1 constant bc array size isnt changing
+      def reverse(array, length)
+        low = 0
+        high = length - 1
+        
+        until low >= high
+          top = array[high]
+          bottom = array[low]
+          array[high] = bottom
+          arry[low] = top
+          high -= 1
+          low += 1
+        end
+        return array
+      end
+      
+      # For an array sorted in ascending order, searches for 'value_to_find'.
+      # Returns true if found, false otherwise.
+      # Time complexity: #binary search/ OofLogN always bc binary. 
+      # Space complexity: Oof1 bc size isnt changing
+      def binary_search(array, length, value_to_find)
+        low = 0
+        high = array.length - 1
+        while low <= high
+          mid = (low + high)/2
+          if (array[mid] > value_to_find
+            high = mid - 1
+          elsif array[mid] < value_to_find
+            low = mid + 1
+          else
+            return mid
+          end
+        end
+        
+        if array[low] == value_to_find
+          return low
```

Comment:
The method is stated above to return true if it find the element. 

```suggestion
          return true
```